src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1852,10 +1878,113 @@ ucs_status_t uct_ib_mlx5_devx_mem_attach(uct_md_h uct_md,
     return status;
 }
 
+static ucs_status_t
+uct_ib_mlx5_devx_device_mem_alloc(uct_md_h uct_md, size_t *length_p,
+                                  void **address_p, ucs_memory_type_t mem_type,
+                                  unsigned flags, const char *alloc_name,
+                                  uct_mem_h *memh_p)
+{
+#if HAVE_IBV_DM && HAVE_DECL_IBV_IMPORT_DM
+    uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
+    uct_md_mem_reg_params_t reg_params = {
+        .field_mask = 0
+    };
+
+    /* Align the allocation to a potential use of registration cache */
+    size_t length = ucs_align_up_pow2(*length_p, md->alloc_align);
+    struct ibv_alloc_dm_attr dm_attr;
+    struct ibv_dm *dm;
+    uct_ib_mlx5_devx_mem_t *memh;
+    void *address;
+    ucs_status_t status;
+    unsigned mem_flags = UCT_MD_MEM_ACCESS_REMOTE_GET |
+                         UCT_MD_MEM_ACCESS_REMOTE_PUT |
+                         UCT_MD_MEM_ACCESS_REMOTE_ATOMIC;
+
+    if (mem_type != UCS_MEMORY_TYPE_RDMA) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (length > md->dev.dev_attr.max_dm_size) {
+        ucs_warn("Tried allocated %ld while max_dm_size is %ld", length,
+                 md->dev.dev_attr.max_dm_size);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    /* Allocate device memory */
+    dm_attr.length    = length;
+    dm_attr.comp_mask = 0;
+
+    status = uct_ib_memh_alloc(md, length, mem_flags,
+                               sizeof(uct_ib_mlx5_devx_mem_t),
+                               sizeof(struct ibv_mr), (uct_ib_mem_t**)&memh);
+    if (status != UCS_OK) {
+        uct_ib_md_log_mem_reg_error(md, reg_params.flags,
+                                    "failed to allocate memory handle");
+        return UCS_ERR_NO_MEMORY;
+    }
+    dm = UCS_PROFILE_CALL(ibv_alloc_dm, md->dev.ibv_context, &dm_attr);
+
+    if (dm == NULL) {
+        ucs_debug("ibv_alloc_dm(dev=%s, length=%zu) failed: %m",
+                  uct_ib_device_name(&md->dev), dm_attr.length);
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    /* Reserve non-accessible address range of the same length as the allocated dm.  */
+    address = mmap(NULL, length, PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    ucs_info("allocated device memory %p..%p on %s", address,
+             UCS_PTR_BYTE_OFFSET(address, length),
+             uct_ib_device_name(&md->dev));
```

Comment:
remove: it should be ucs_debug, and it's also misleading because address is not device memory address but a fake VA we got from mmap

Reply:
So to remove or to change to `ucs_debug` and to "%s: allocated device memory **virtual address** %p..%p"?

Reply:
`ucs_debug("%s: allocated device memory %p length %zu, attached at virtual address %p", device_name, dm, length, address);`